lcls-twincat-motion/Library/POUs/Motion/FB_MotionHoming.TcPOU

```diff
@@ -0,0 +1,245 @@
+            ELSIF fbJog.Error THEN
+                fbJog(
+                    Axis:=stMotionStage.Axis,
+                    JogForward:=FALSE,
+                    JogBackwards:=FALSE);
+                nErrCount := nErrCount + 1;
+                IF nErrCount >= 3 THEN
+                    nHomeStateMachine := ERROR;
+                END_IF
+            END_IF
+        // Move backward until we find the home signal or reach end of travel
+        CHECK_BWD:
+            fbSetPos(
+                Axis:=stMotionStage.Axis,
+                Execute:=TRUE,
+                Position:=BWD_START);
+            fbJog(
+                Axis:=stMotionStage.Axis,
+                JogForward:=FALSE,
+                JogBackwards:=stMotionStage.bLimitBackwardEnable AND NOT bATHome,
+                Mode:=E_JogMode.MC_JOGMODE_CONTINOUS,
+                Velocity:=stMotionStage.stAxisParameters.fRefVeloSearch);
+            IF NOT fbJog.JogBackwards THEN
+                nHomeStateMachine := NEXT_MOVE;
+            ELSIF fbJog.Error THEN
+                fbJog(
+                    Axis:=stMotionStage.Axis,
+                    JogForward:=FALSE,
+                    JogBackwards:=FALSE);
+                nErrCount := nErrCount + 1;
+                IF nErrCount >= 3 THEN
+                    nHomeStateMachine := ERROR;
+                END_IF
+            END_IF
+        // Set position to get within soft lims, move slowly off signal
+        FINAL_MOVE:
+            fbSetPos(
+                Axis:=stMotionStage.Axis,
+                Execute:=TRUE,
+                Position:=stMotionStage.fHomePosition);
+            IF bAtHome THEN
+                fbJog(
+                    Axis:=stMotionStage.Axis,
+                    JogForward:=NOT bFirstDirection,
+                    JogBackwards:=bFirstDirection,
+                    Mode:=E_JogMode.MC_JOGMODE_CONTINOUS,
+                    Velocity:=stMotionStage.stAxisParameters.fRefVeloSync);
+            ELSIF fbJog.Error THEN
+                fbJog(
+                    Axis:=stMotionStage.Axis,
+                    JogForward:=FALSE,
+                    JogBackwards:=FALSE);
+                nErrCount := nErrCount + 1;
+                IF nErrCount >= 3 THEN
+                    nHomeStateMachine := ERROR;
+                END_IF
+            ELSE
+                fbJog(
+                    Axis:=stMotionStage.Axis,
+                    JogForward:=FALSE,
+                    JogBackwards:=FALSE);
+                fbSetPos(
+                    Axis:=stMotionStage.Axis,
+                    Execute:=FALSE);
+                nHomeStateMachine := FINAL_SETPOS;
+            END_IF
+        FINAL_SETPOS:
+            fbSetPos(
+                Axis:=stMotionStage.Axis,
+                Execute:=TRUE,
+                Position:=stMotionStage.fHomePosition);
+            nHomeStateMachine := IDLE;
+            bBusy := FALSE;
+            bDone := TRUE;
+        ERROR:
```

Comment:
In testing I always ended up at a "normal-looking" position after any error handling, since the error state machine hops over to the final position set, which gives us the "home" position. The motor will not be allowed to move in this case because it shows as `bError`.